python3.9 causes assertion error

ran into the following issue using the add-trailing-comma hook on macOS with python3.9 (works fine with `<= 3.8`). `.pre-commit-config.yaml` is the one from [pre-commit](https://github.com/pre-commit/pre-commit/blob/master/.pre-commit-config.yaml).

i tried reinstalling python3.9 and cleaning the pre-commit cache but nothing changed.

error and system info:

```console
+ sw_vers
ProductName:	macOS
ProductVersion:	11.1
BuildVersion:	20C69
```

```console
+ python --version
Python 3.9.1
```

```console
+ pre-commit --version
pre-commit 2.9.3
```

```console
+ pre-commit run add-trailing-comma --all-files
Add trailing commas......................................................Failed
- hook id: add-trailing-comma
- exit code: 1

Traceback (most recent call last):
  File "/Users/paulhfischer/.cache/pre-commit/repothj9r_n4/py_env-python3.9/bin/add-trailing-comma", line 8, in <module>
    sys.exit(main())
  File "/Users/paulhfischer/.cache/pre-commit/repothj9r_n4/py_env-python3.9/lib/python3.9/site-packages/add_trailing_comma.py", line 498, in main
    ret |= fix_file(filename, args)
  File "/Users/paulhfischer/.cache/pre-commit/repothj9r_n4/py_env-python3.9/lib/python3.9/site-packages/add_trailing_comma.py", line 470, in fix_file
    contents_text = _fix_src(contents_text, args.py35_plus, args.py36_plus)
  File "/Users/paulhfischer/.cache/pre-commit/repothj9r_n4/py_env-python3.9/lib/python3.9/site-packages/add_trailing_comma.py", line 401, in _fix_src
    tokens, _find_call(call, i, tokens),
  File "/Users/paulhfischer/.cache/pre-commit/repothj9r_n4/py_env-python3.9/lib/python3.9/site-packages/add_trailing_comma.py", line 239, in _find_call
    raise AssertionError('Past end?')
AssertionError: Past end?
```

--- a/add_trailing_comma.py
+++ b/add_trailing_comma.py
@@ -50,7 +50,7 @@ def _to_offset(node: ast.AST) -> Offset:
         candidate = candidates.pop()
         if hasattr(candidate, 'lineno'):
             return Offset(candidate.lineno, candidate.col_offset)
-        elif hasattr(candidate, '_fields'):
+        elif hasattr(candidate, '_fields'):  # pragma: no cover (<py39)
             for field in reversed(candidate._fields):
                 candidates.append(getattr(candidate, field))
     else:
@@ -66,6 +66,19 @@ class FindNodes(ast.NodeVisitor):
         self.tuples: Dict[Offset, ast.Tuple] = {}
         self.imports: Set[Offset] = set()
         self.classes: Dict[Offset, Node] = {}
+
+        # https://bugs.python.org/issue42806
+        # the tokenizer doesn't sub-tokenize in f-string substitutions so
+        # normally calls in there never match, however python3.9's new parser
+        # is buggy
+        self._in_fstring = False
+
+    def visit_FormattedValue(self, node: ast.FormattedValue) -> None:
+        orig, self._in_fstring = self._in_fstring, True
+        try:
+            self.generic_visit(node)
+        finally:
+            self._in_fstring = orig
 
     def _visit_literal(self, node: ast.expr, key: str = 'elts') -> None:
         if getattr(node, key):
@@ -109,7 +122,7 @@ class FindNodes(ast.NodeVisitor):
             len(argnodes) == 1 and isinstance(argnodes[0], ast.GeneratorExp)
         )
 
-        if arg_offsets and not only_a_generator:
+        if arg_offsets and not only_a_generator and not self._in_fstring:
             key = _to_offset(node)
             self.calls[key].append(Node(has_starargs, arg_offsets))
 

--- a/azure-pipelines.yml
+++ b/azure-pipelines.yml
@@ -10,7 +10,7 @@ resources:
       type: github
       endpoint: github
       name: asottile/azure-pipeline-templates
-      ref: refs/tags/v1.0.1
+      ref: refs/tags/v2.1.0
 
 jobs:
 - template: job--python-tox.yml@asottile
@@ -19,5 +19,5 @@ jobs:
     os: windows
 - template: job--python-tox.yml@asottile
   parameters:
-    toxenvs: [pypy3, py36, py37, py38]
+    toxenvs: [pypy3, py36, py37, py38, py39]
     os: linux

--- a/tests/add_trailing_comma_test.py
+++ b/tests/add_trailing_comma_test.py
@@ -41,6 +41,13 @@ from add_trailing_comma import main
         '(\n'
         '    a\n'
         ').f(b)',
+        pytest.param(
+            'x = (\n'
+            '    f" {test(t)}"\n'
+            ')\n',
+
+            id='regression test for #106',
+        ),
     ),
 )
 def test_fix_calls_noops(src):

--- a/tox.ini
+++ b/tox.ini
@@ -1,5 +1,5 @@
 [tox]
-envlist = py36,py37,py38,pypy3,pre-commit
+envlist = py36,py37,py38,py39,pypy3,pre-commit
 
 [testenv]
 deps = -rrequirements-dev.txt

